src/types/store.js

```diff
@@ -19,7 +12,9 @@ import type { State as StateRef } from './reducers';
 export type Action = ActionsRef;
 export type State = StateRef;
 
-export type GetState = ReduxGetState<State>;
-export type ThunkAction<Result> = ReduxThunkAction<State, Action, Result>;
-export type Store = ReduxStore<State, Action>;
-export type Dispatch = ReduxDispatch<State, Action>;
+type ThunkDispatch = <R>(action: ThunkAction<R>) => R;
+type PlainDispatch = (action: Action) => Action;
+export type GetState = () => State;
+export type ThunkAction<R> = (dispatch: Dispatch, GetState) => R;
+export type Dispatch = PlainDispatch & ThunkDispatch;
```

Comment:
Is this supposed to be an `|` instead of `&`?

Reply:
Nope :) dispatch takes both an action or a thunk action

Reply:
I added a comment explaining the reasoning. Note this is nearly identical to what was before, but in the flow-typed definition :) (written by yourself IIRC ;) )

Reply:
Hmm... I can't [get this to work in a simple example](https://flow.org/try/#0C4TwDgpgBAggxsAlgewHZQLxQPICMBWECA3AFCiRQDKwAhsNFnoSaeeNACoAWArqgGsAIogDOYenG6YoAHgBKAPgAUtBClQAuKD34D4SNAsUBKTIqjyyFaAAUANrUSoR4ydKyr1abQY1mMCz80MggADzBkACdgKBsoAHEIYBp6RihlAItUhlCI6Nj43UFg1GMZZQATMQlgKW1XWqkAGkTknIgsyzzImLiOKEb3GQcnFxrhgB8dPkEhuu4yKGWV5dIAM35vdABbEFHneakvQy1YbZNfbagAb1JlqOTeKPQbgF8yN-uV0jg0UVi1TcCwaEwWMj2B3GwKkZA2W1OUD2xX020yt2+j2Az3Qm1QaKBTW4oJh3FaAHN2nQGNokilqZ0MassTioAByWhQAFRZzktlLKBvVakL5467I2bCMFSDLASWlbQo0qyABUpiZUD+qABUEp9LSFS6yneJgFLJecXlaMhjkO0rJuqpaVNItIQA) with an intersection `&` as opposed to a union `|`. I really don't understand how this works. There aren't any examples of intersections with functions on the flow website.

Reply:
look at [this example](https://flow.org/try/#0C4TwDgpgBAggxsAlgewHZQLxQN5VJALigGdgAnRVAcygF8BuAKH2gGVgBDYaLXAM2TIipCtTpNm4aABUAFgFdUAawAiiYmC5xZmKAB4ASgD4AFBwQpUROYqXwkaQ0YCUmI1ANMWUAAoAbDko1DS0dLDMLNCJ7S1cMdxi0JggADzBkMmA8KSgAcQhgdi4eKBM49yLuZLSMrO8bZUTUJ10TABN1TWBtImCu7QAaPILKiHKPavTM7MgoPtDdf0DUee6dADIoBtVO0KYoA8Y+RUj0blJ23bXeq+1XbEYDjpC1k1wWIgByACMOMk+6M59lBnv1ZCZgApGqcys4AHSQiCoEzfQR+NxQOBoYjIPwQOF+ZBUFFo5xAx4g27gxQsNrbJqwuECZBMWiMCmHY6oU54KF2GHOax8pp6HxkZAAW3UED0ACojO4HgdlVAyAV5GR0KCFvEcBzlQBIbWvd5SL6-f6A4EqqBGqkQ4UC8k25UAeldlJe2jeUGZ5r+ANoQKg7qgAGkAPJDKiCNpQACE+oOBrVwA16DFkulcLVOL8ADcIBCyPIxtblWy2S6jicHOgaVI6Y662U9S7U+nPWCMUqbXaveDTYQoD8A1ak8njd7IbYGWTyyrQ1PB77BP7LUH6CGPZHo7GExOU+rNThV0IRxbAwu6IxaEA)

The difference is that you don't create dispatch yourself, you get it passed by redux. It happens that `dispatch` supports these 2 parameters thanks to the middlewares.

So in this example I use a function where `dispatch` is passed with the right type. (this is a common way to create example with try flow :) ).

I think this works because redux exposes a specific flow interface we use, but internally it likely has another flow mechanism. And all this works because we don't use redux' internal interface, we use its public interface exposed by flow-typed.

Reply:
to be clear, I don't exactly know how to produce a real dispatch function that would obey this interface -- but all we need to know is that the passed `dispatch` obey this interface, we don't care how it's done.